Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs

```diff
@@ -745,15 +719,25 @@ public void RaiseFocusEnter(IMixedRealityPointer pointer, GameObject focusedObje
                 delegate (IMixedRealityFocusHandler handler, BaseEventData eventData)
                 {
                     var casted = ExecuteEvents.ValidateEventData<FocusEventData>(eventData);
-                    handler.OnFocusEnter(casted);
+                    if (handler.ReceiveAllFocusChanges)
```

Comment:
I feel like this logic doesn't belong here. We have three classes (FocusManager, InputSystem and FocusHander) cooperating to decide if the event should be raised. We should instead raise the event always and leave it to the FocusHandler to decide what to do with it.

Reply:
that's what @davidkline-ms wasn't sure with, he mentions that further up the discussion.

Reply:
> Changing to send FocusEnter and FocusExit for each pointe, while giving controls that wish to track each pointer individually more information, it would require all controls to have their own tracking logic to avoid potential odd UX and state behaviors.

I think that's the relevant bit. Correct me if I'm wrong but I think the logic needed to implement the current behavior is something like this:

```
int count = 0;
Pointer pointer;

OnFocusEnter(event)
{
	count++;
	if (count == 1)
	{
		pointer = event.pointer;
	}
}

OnFocusExit(event)
{
	count--;
	if (event.pointer == pointer)
	{
		pointer = null;
	}		
} 
```
It doesn't seem too bad but we could have that in BaseFocusHandler, something like:
```
class BaseFocusHandler
{
	bool receiveAllFocusEvents = false;
	
	abstract HandleFocusEnter();
	
	sealed OnFocusEnter(event)
	{
		if (!receiveAllFocusEvents)
		{
			count++;
			if (count == 1)
			{
				HandleFocusEnter(event);
			}			
		}
	}
	
	// ...	
}
```
What do you think?

Reply:
this is about what this does already, with the tiny convenience extra that you don't have to check for yourself, that's what the InputManager does.
If handler.ReceiveAllFocusChanges is false you just won't receive but the single ones. if it's true you will.
My feeling is that's cleaner.

Reply:
I have to disagree with you there. Making this a pure UI concern, i.e. private to the component handling focus, provides better encapsulation which makes the system as a whole easier to understand, use and change. 

With the current solution in this PR, we rely on implementations of the input system interface checking the property in the focus handler before raising focus enter/left events. This is not enforced in any way by the interface definitions, so you will only realize you got it wrong when it fails at runtime instead of at compile time. If UI components take responsability for it instead, it is more likely they will work fine with a different input system implementation.

Reply:
I agree with @luis-valverde-ms this should be something the UI should be concerned about, not the focus provider.

Reply:
@luis-valverde-ms it is enforced by the interface IMixedRealityFocusHandler. you make that Choice when implementing it.

Reply:
@StephenHodgson just to be clear about this. I would say the majority of cases requires just a single enter and exit. And the receivers do require an interface to receive them.
This means most of the time you write a counter every time you implement this interface if you want just the first and the last call. Or you start with a base class that does it, introducing virtual calls for receiving just single calls.
With this PR it's merely letting ReceiveAllFocusChanges return false and you're good to go. And it makes it very clear during the implementation what calls will be received, all or one.
My hand up for the PR.

Reply:
@Alexees, totally agree with you in that we should make the most common use case as easy as possible to implement. Not at the cost of introducing that dependency in the Input System though. 

I feel that we already have the mechanism to allow UI implementers to choose between first/all focus events in the shape of the Focus and FocusChanged handler interfaces. If there are no strong objections, I'm going to go ahead and change Interactable to keep track of the pointers using FocusChanged so we can close #3577.

Reply:
PR ready: #3670

Reply:
Am I correct that the FocusProvider is a service and should therefore not be concerned with limiting focus enter and exit calls, which it currently does?

Reply:
> If there are no strong objections, I'm going to go ahead and change Interactable to keep track of the pointers using FocusChanged so we can close

@luis-valverde-ms if we leave the FocusProvider as is, there is no need to count pointers in Interactables, because FocusProvider already does and fires enter and exit just once.

Reply:
I'm no Interactable expert but it seems that it is meant to keep a list of all the pointers focusing it and expose it via the `Focusers` property.

Reply:
Ok, so to sum it up:
- The Focus Provider counts pointers and makes sure enter and exit are called just once
- PreFocusChanged and FocusChanged are called for all changes
- Interactables count pointers too, so there's information about which pointers are focusing

To be honest, I don't like having 4 different events for focusing and two or more places where pointers are counted, but that's probably just me.
I'd rather like to see a solution with only 2 events and one place for counting, referenced when needed. I mean all other events do not have pre and post events. PrePointerClick anyone?